node/src/components/consensus/highway_core/state.rs

```diff
@@ -211,11 +216,47 @@ impl<C: Context> State<C> {
         self.opt_fault(idx).and_then(Fault::evidence)
     }
 
+    /// Returns endorsements for `vote`, if any.
+    pub(crate) fn opt_endorsements(&self, vote: &C::Hash) -> Option<Vec<Endorsement<C>>> {
+        self.endorsement.get(vote).cloned()
+    }
+
     /// Returns whether evidence against validator nr. `idx` is known.
     pub(crate) fn has_evidence(&self, idx: ValidatorIndex) -> bool {
         self.opt_evidence(idx).is_some()
     }
 
+    /// Returns whether we have all endorsements for `vote`.
+    pub(crate) fn has_all_endorsements<'a, I: IntoIterator<Item = &'a ValidatorIndex>>(
+        &self,
+        vote: &C::Hash,
+        v_ids: I,
+    ) -> bool {
+        self.endorsement
+            .get(vote)
+            .map(|v| {
+                v_ids
+                    .into_iter()
+                    .all(|v_id| v.get(v_id.0 as usize).is_some())
+            })
+            .unwrap_or_else(|| false)
+    }
+
+    /// Returns whether we have seen enough endorsements for the vote.
+    pub(crate) fn is_endorsed(&self, hash: &C::Hash) -> bool {
```

Comment:
I wonder whether it would be cleaner and more efficient to better distinguish between complete and incomplete `Endorsements` (≥ ⅔), by separating the field `endorsement` (should that be plural anyway?) into two fields — collecting them in `incomplete_endorsements`, and moving them to `endorsements` once they're enough —, or maybe even with a separate type?
It would make things easier if a `Vertex::Endorsements` always meant that the dependency on endorsements for that vote is satisfied, i.e. that it's ≥ ⅔.

Reply:
I will give a shot at separating `incomplete_endorsements` from `endorsements` but I am unsure about `Vertex::Endorsements` representing set of vote endorsements that have ≥ ⅔ stake. Reason is that if we follow the scheme you proposed in the email thread (where we send endorsements for every vote that is "endorse-able") then the endorsement sent there won't represent the "completely endorsed" so that's yet another `Vertex` variant:
1. Single endorsement.
2. Complete set of endorsements.

How about I use the first part of your suggestion and see how it works out?

Reply:
Done.